bpf/libbpf/bpf_tracing.h

```diff
@@ -0,0 +1,388 @@
+#define BPF_KRETPROBE_READ_RET_IP	BPF_KPROBE_READ_RET_IP
+
+#else
+
+#define BPF_KPROBE_READ_RET_IP(ip, ctx)                                        \
+	({                                                                     \
+		bpf_probe_read_kernel(&(ip), sizeof(ip),                       \
+				      (void *)PT_REGS_RET(ctx));               \
+	})
+#define BPF_KRETPROBE_READ_RET_IP(ip, ctx)                                     \
+	({                                                                     \
+		bpf_probe_read_kernel(&(ip), sizeof(ip),                       \
+				      (void *)(PT_REGS_FP(ctx) + sizeof(ip))); \
+	})
+
+#endif
+
+#ifndef PT_REGS_PARM1_SYSCALL
+#define PT_REGS_PARM1_SYSCALL(x) PT_REGS_PARM1(x)
+#endif
+#define PT_REGS_PARM2_SYSCALL(x) PT_REGS_PARM2(x)
+#define PT_REGS_PARM3_SYSCALL(x) PT_REGS_PARM3(x)
+#ifndef PT_REGS_PARM4_SYSCALL
+#define PT_REGS_PARM4_SYSCALL(x) PT_REGS_PARM4(x)
+#endif
+#define PT_REGS_PARM5_SYSCALL(x) PT_REGS_PARM5(x)
+
+#ifndef PT_REGS_PARM1_CORE_SYSCALL
+#define PT_REGS_PARM1_CORE_SYSCALL(x) PT_REGS_PARM1_CORE(x)
+#endif
+#define PT_REGS_PARM2_CORE_SYSCALL(x) PT_REGS_PARM2_CORE(x)
+#define PT_REGS_PARM3_CORE_SYSCALL(x) PT_REGS_PARM3_CORE(x)
+#ifndef PT_REGS_PARM4_CORE_SYSCALL
+#define PT_REGS_PARM4_CORE_SYSCALL(x) PT_REGS_PARM4_CORE(x)
+#endif
+#define PT_REGS_PARM5_CORE_SYSCALL(x) PT_REGS_PARM5_CORE(x)
+
+#else /* defined(bpf_target_defined) */
+
+#define PT_REGS_PARM1(x) 0
+#define PT_REGS_PARM2(x) 0
+#define PT_REGS_PARM3(x) 0
+#define PT_REGS_PARM4(x) 0
+#define PT_REGS_PARM5(x) 0
+#define PT_REGS_RET(x)	 0
+#define PT_REGS_FP(x)	 0
+#define PT_REGS_RC(x)	 0
+#define PT_REGS_SP(x)	 0
+#define PT_REGS_IP(x)	 0
+
+#define PT_REGS_PARM1_CORE(x) 0
+#define PT_REGS_PARM2_CORE(x) 0
+#define PT_REGS_PARM3_CORE(x) 0
+#define PT_REGS_PARM4_CORE(x) 0
+#define PT_REGS_PARM5_CORE(x) 0
+#define PT_REGS_RET_CORE(x)   0
+#define PT_REGS_FP_CORE(x)    0
+#define PT_REGS_RC_CORE(x)    0
+#define PT_REGS_SP_CORE(x)    0
+#define PT_REGS_IP_CORE(x)    0
+
+#define BPF_KPROBE_READ_RET_IP(ip, ctx)	   0
+#define BPF_KRETPROBE_READ_RET_IP(ip, ctx) 0
+
+#define PT_REGS_PARM1_SYSCALL(x) 0
+#define PT_REGS_PARM2_SYSCALL(x) 0
+#define PT_REGS_PARM3_SYSCALL(x) 0
+#define PT_REGS_PARM4_SYSCALL(x) 0
+#define PT_REGS_PARM5_SYSCALL(x) 0
+
+#define PT_REGS_PARM1_CORE_SYSCALL(x) 0
+#define PT_REGS_PARM2_CORE_SYSCALL(x) 0
+#define PT_REGS_PARM3_CORE_SYSCALL(x) 0
+#define PT_REGS_PARM4_CORE_SYSCALL(x) 0
+#define PT_REGS_PARM5_CORE_SYSCALL(x) 0
+
+#endif /* defined(bpf_target_defined) */
+
+/*
+ * When invoked from a syscall handler kprobe, returns a pointer to a
+ * struct pt_regs containing syscall arguments and suitable for passing to
+ * PT_REGS_PARMn_SYSCALL() and PT_REGS_PARMn_CORE_SYSCALL().
+ */
+#ifndef PT_REGS_SYSCALL_REGS
+/* By default, assume that the arch selects ARCH_HAS_SYSCALL_WRAPPER. */
+#define PT_REGS_SYSCALL_REGS(ctx) ((struct pt_regs *)PT_REGS_PARM1(ctx))
+#endif
+
+#ifndef ___bpf_concat
```

Comment:
Thanks for your review! Yes, it's better to include `bpf_tracing.h` in a single commit, I'll revise it.

___

In fact I only include the minimum necessary definitions in the first commit.

It's tpapagian's advice to include the whole file directly from the libbpf. I also think it's good because it can ease developments in the future, so that we don't need to compare the header with libbpf if more features are need.

Then the second commit is submit.